api/src/main/java/jakarta/persistence/criteria/CriteriaBuilder.java

```diff
@@ -902,6 +906,69 @@ public interface CriteriaBuilder {
      */
     <T> Expression<T> literal(T value);
 
+    /**
+     * Create a number expression for a literal.
+     * @param value  value represented by the expression
+     * @return expression literal
+     * @throws IllegalArgumentException if value is null
+     * @since 3.2
+     */
+    <T extends Number> NumberExpression<T> literal(T value);
+
+    /**
+     * Create a boolean expression for a literal.
+     * @param value  value represented by the expression
+     * @return expression literal
+     * @throws IllegalArgumentException if value is null
+     * @since 3.2
+     */
+    BooleanExpression literal(Boolean value);
```

Comment:
Don't you need an `asBoolean(Expression<Boolean>)` as well? (Also for String, Number, Temporal, LocalDate, LocalDateTime). Becomes more important if you decide not to add the covariant overrides, so one can still do: `cb.asNumber(entity.overrideValue.coalesce(parent.defaultValue)).avg()`.

Alternatively those can be implemented as `asBoolean()` on `Expression`.